scripts/generators/intermediate_files.py

```diff
@@ -1,8 +1,113 @@
+import copy
+
+from schema import visitor
 from generators import ecs_helpers
 from os.path import join
 
 
-def generate(ecs_nested, ecs_flat, out_dir):
+def generate(fields, out_dir, default_dirs):
     ecs_helpers.make_dirs(join(out_dir, 'ecs'))
-    ecs_helpers.yaml_dump(join(out_dir, 'ecs/ecs_flat.yml'), ecs_flat)
-    ecs_helpers.yaml_dump(join(out_dir, 'ecs/ecs_nested.yml'), ecs_nested)
+
+    # Should only be used for debugging ECS development
+    if default_dirs:
+        ecs_helpers.yaml_dump(join(out_dir, 'ecs/ecs.yml'), fields)
+
+    flat = generate_flat_fields(fields)
+    nested = generate_nested_fields(fields)
+
+    ecs_helpers.yaml_dump(join(out_dir, 'ecs/ecs_flat.yml'), flat)
+    ecs_helpers.yaml_dump(join(out_dir, 'ecs/ecs_nested.yml'), nested)
+    return nested, flat
+
+
+def generate_flat_fields(fields):
+    '''Generate ecs_flat.yml'''
+    filtered = remove_non_root_reusables(fields)
+    flattened = {}
+    visitor.visit_fields_with_memo(filtered, accumulate_flat_field, flattened)
+    return flattened
+
+
+def accumulate_flat_field(details, memo):
+    '''Visitor function that accumulates all flat field details in the memo dict'''
+    if 'schema_details' in details or ecs_helpers.is_intermediate(details):
+        return
+    field_details = copy.deepcopy(details['field_details'])
+    remove_internal_attributes(field_details)
+
+    # TODO Remove workaround field until it's removed as a later PR
+    field_details.pop('ctx_name')
+
+    flat_name = field_details['flat_name']
+    memo[flat_name] = field_details
```

Comment:
Yeah this is very close to the last remaining major issue with this PR (the tracing fields getting mangled). I definitely see this `ctx_name` as a temporary hack, and I'm not letting it appear in either ecs_flat.yml nor ecs_nested.yml.

I'll see if I can get rid of it in this PR.

I'm pretty annoyed by the fact that I ended up with 3 implementations of the visitor pattern, by the way 🤦 . I've actually been considering getting rid of the `path` approach completely, however. Initially I thought it would simplify calculating the final field names in the context of performing field reuse.

But it turns out that this breaks when we have a field set that's `root=true` but also has nested fields. The attribute `root=true` was initially conceived for base fields, but now it's also used for the "tracing" field set, with its fields `trace.id` and `transaction.id`... In the path-based approach I currently don't have the info I need to determine if I therefore need to chop off the first part of the path (`[base, @timestamp]` => `@timestamp`) or I should keep it around.

I'll consider your input when looking at fixing all this.

Reply:
Good observation on the two functions being almost the same. With my recent fix nesting all fields as their full field name instead of their contextual name, the two functions were now exact duplicates, so I've removed one of them.

Reply:
I haven't merged the two visitor implementations with memo and with path. Their usage is similar enough that it could be done (as opposed to visit_fields, with specific handling for schemas).

I don't see a huge benefit of doing it in this PR. I'd rather merge this sooner than later, because there's 3 subsequent PRs to this I'm waiting to open. I'll make a note to check it out in one of the follow-up PRs (the one that introduces additional cleanups and small changes), deal?